README.md
Sicromoft---Project-3-Data-Mining
=================================

Project 3 Data-Mining for college

This was a project for college.
The theme of the project was "Big data". The goal of the project was to search for data, mine into databases and analyze
the data.


changelog.txt
Week 4:
- Sentiment proeven uitgevoerd

Week 5:
- Onderzoek database aanmaken in R
- Sentiment scripts toegevoegd in applicatie

Week 6:
- Onderzoek database in R
- Database opgezet met package RMySQL
- Steekproeven met package RMySQL

Week 7:
- Implementatie database in R
- Tweets verzamelen en in database opslaan
- Begin ontwerp van applicatie

Week 8:
- Bouwen van applicatie
- Alle geslaagde geteste code in de applicatie gezet
- UI en Server code aangepast
- Extra functies toegevoegd
- Onnodige code en commentaar weggehaald




UI.R
library(shiny)

# Define UI for miles per gallon application
shinyUI(pageWithSidebar(
  
  # Application title
  headerPanel("Sicromoft Twitter Sentiment"),
  
  sidebarPanel(
    headerPanel("Sentiment Twitter 1"),
    selectInput("variable", "Variable:",
                list("Efteling" = "eft", 
                     "Dolfinarium" = "dolf", 
                     "Pathe" = "pathe",
                     "Apenheul" = "aap")),
    
    headerPanel("Sentiment Twitter 2"),
    selectInput("variable1", "Variable:",
                list("Efteling" = "eft2", 
                     "Dolfinarium" = "dolf2", 
                     "Pathe" = "pathe2",
                     "Apenheul" = "aap2")),
    
    headerPanel("Tweet Frequency"),
    selectInput("variable2", "Variable:",
                list("Efteling" = "eft3", 
                     "Dolfinarium" = "dolf3", 
                     "Pathe" = "pathe3",
                     "Apenheul" = "aap3")),
    
    helpText("Note: Getting results may take a while.
             Please be patient."),
    
    submitButton("Update View")
    ),
  
  mainPanel(
    tabsetPanel(
      tabPanel("Sentiment Twitter 1", plotOutput("STPlot1")),
      tabPanel("Sentiment Twitter 2", plotOutput("STPlot2")),
      tabPanel("Tweet Frequency", plotOutput("tweetFreq"))
      )
    )
))

sentiment.R
#' 
#' score.sentiment() implements a very simple algorithm to estimate
#' sentiment, assigning a integer score by subtracting the number 
#' of occurrences of negative words from that of positive words.
#' 
#' @param sentences vector of text to score
#' @param pos.words vector of words of postive sentiment
#' @param neg.words vector of words of negative sentiment
#' @param .progress passed to <code>laply()</code> to control of progress bar.
#' @returnType data.frame
#' @return data.frame of text and corresponding sentiment scores
#' @author Jefrey Breen <jbreen@cambridge.aero>

score.sentiment = function(sentences, pos.words, neg.words, .progress='none')
{
	require(plyr)
	require(stringr)
	
	# we got a vector of sentences. plyr will handle a list or a vector as an "l" for us
	# we want a simple array of scores back, so we use "l" + "a" + "ply" = laply:
	scores = laply(sentences, function(sentence, pos.words, neg.words) {
		
		# clean up sentences with R's regex-driven global substitute, gsub():
		sentence = gsub('[[:punct:]]', '', sentence)
		sentence = gsub('[[:cntrl:]]', '', sentence)
		sentence = gsub('\\d+', '', sentence)
		# and convert to lower case:
		sentence = tolower(sentence)

		# split into words. str_split is in the stringr package
		word.list = str_split(sentence, '\\s+')
		# sometimes a list() is one level of hierarchy too much
		words = unlist(word.list)

		# compare our words to the dictionaries of positive & negative terms
		pos.matches = match(words, pos.words)
		neg.matches = match(words, neg.words)
	
		# match() returns the position of the matched term or NA
		# we just want a TRUE/FALSE:
		pos.matches = !is.na(pos.matches)
		neg.matches = !is.na(neg.matches)

		# and conveniently enough, TRUE/FALSE will be treated as 1/0 by sum():
		score = sum(pos.matches) - sum(neg.matches)

		return(score)
	}, pos.words, neg.words, .progress=.progress )

	scores.df = data.frame(score=scores, text=sentences)
	return(scores.df)
}


ggplotTwatter.R
library(twitteR)
library(ggplot2)

load(file="twitter_authentication.Rdata")
registerTwitterOAuth(cred)

tweets <- list()
dates <- paste("2013-04-",7:11,sep="") # need to go to 18th to catch tweets from 17th
for (i in 2:length(dates)) {
  print(paste(dates[i-1], dates[i]))
  tweets <- c(tweets, searchTwitter("#dolfinarium", since=dates[i-2], until=dates[i], n=150))
}

# Convert the list to a data frame
#tweets <- twListToDF(tweets)
tweets <- do.call("rbind", lapply(tweets, as.data.frame))
write.csv(tweets, file = "eftelingFreq.csv")
#tweets <- unique(tweets)

# To ensure accuracy, make sure that there were no more than 1500 tweets in a single day.
# If there are 1500 on any single day, then you're truncating that day's tweets, and you'll
# need to try to get ROAuth (below) working.
tweets$date <- format(tweets$created, format="%Y-%m-%d")
table(tweets$date)

# Plot the frequency of tweets over time in two hour windows
# Modified from http://michaelbommarito.com/2011/03/12/a-quick-look-at-march11-saudi-tweets/
minutes <- 120
ggplot(data=tweets, aes(x=created)) + 
  geom_bar(aes(fill=..count..), binwidth=60*minutes) + 
  scale_x_datetime("Date") + 
  scale_y_continuous("Frequency") +
  ggtitle("#Dolfinarium Tweet Frequency March 05-11")


Server.R
library(shiny)
library(twitteR)
library(RMySQL)
library(ROAuth)
library(ggplot2)

#Load authentication files for twitter
load(file="twitter_authentication.Rdata")
registerTwitterOAuth(cred)

tweetsEfteling <- function()
{
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  eftelingResultText <- dbSendQuery(doConnectToDB, "select text from tweetsEfteling")
  showResultEftelingText <- fetch(eftelingResultText)
  onComplete <- dbHasCompleted(eftelingResultText)
  dbClearResult(eftelingResultText)
  dbDisconnect(doConnectToDB)
  
  getText <- showResultEftelingText$text
  getText <- iconv(as.character(getText), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  efteling.score <- suppressMessages(score.sentiment(getText, pos.words, neg.words))
  
  hist(main="Sentiment Efteling (march 30th 2013)", efteling.score$score, breaks = (-10:10 + 0.7))
}

tweetsDolfinarium <- function()
{
  
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  dolfinariumResultText <- dbSendQuery(doConnectToDB, "select text from tweetsDolfinarium")
  showResultDolfinariumText <- fetch(dolfinariumResultText)
  onComplete <- dbHasCompleted(dolfinariumResultText)
  dbClearResult(dolfinariumResultText)
  dbDisconnect(doConnectToDB)
  
  
  getText1 <- showResultDolfinariumText$text
  getText1 <- iconv(as.character(getText1), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  dolfinarium.score <- suppressMessages(score.sentiment(getText1, pos.words, neg.words))
  
  hist(main="Sentiment Dolfinarium (march 30th 2013)", dolfinarium.score$score, breaks = (-10:10 + 0.7))
}

tweetsPathe <- function()
{
  
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  patheResultText <- dbSendQuery(doConnectToDB, "select text from tweetsPathe")
  showResultPatheText <- fetch(patheResultText)
  onComplete <- dbHasCompleted(patheResultText)
  dbClearResult(patheResultText)
  dbDisconnect(doConnectToDB)
  
  
  getText2 <- showResultPatheText$text
  getText2 <- iconv(as.character(getText2), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  pathe.score <- suppressMessages(score.sentiment(getText2, pos.words, neg.words))
  
  hist(main="Sentiment Pathe (march 30th 2013)", pathe.score$score, breaks = (-10:10 + 0.7))
}

tweetsApenheul <- function()
{
  
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  apenheulResultText <- dbSendQuery(doConnectToDB, "select text from tweetsApenheul")
  showResultApenheulText <- fetch(apenheulResultText)
  onComplete <- dbHasCompleted(apenheulResultText)
  dbClearResult(apenheulResultText)
  dbDisconnect(doConnectToDB)
  
  
  getText3 <- showResultApenheulText$text
  getText3 <- iconv(as.character(getText3), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  apenheul.score <- suppressMessages(score.sentiment(getText3, pos.words, neg.words))
  
  hist(main="Sentiment Apenheul (march 30th 2013)", apenheul.score$score, breaks = (-10:10 + 0.7))
}

##################################

tweetsEfteling1 <- function()
{
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  eftelingResultText <- dbSendQuery(doConnectToDB, "select text from tweetsEfteling2")
  showResultEftelingText <- fetch(eftelingResultText)
  onComplete <- dbHasCompleted(eftelingResultText)
  dbClearResult(eftelingResultText)
  dbDisconnect(doConnectToDB)
  
  
  getText <- showResultEftelingText$text
  getText <- iconv(as.character(getText), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  efteling.score <- suppressMessages(score.sentiment(getText, pos.words, neg.words))
  
  hist(main="Sentiment Efteling (april 4th 2013)", efteling.score$score, breaks = (-10:10 + 0.7))
}

tweetsDolfinarium1 <- function()
{
  
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  dolfinariumResultText <- dbSendQuery(doConnectToDB, "select text from tweetsDolfinarium2")
  showResultDolfinariumText <- fetch(dolfinariumResultText)
  onComplete <- dbHasCompleted(dolfinariumResultText)
  dbClearResult(dolfinariumResultText)
  dbDisconnect(doConnectToDB)
  
  
  getText1 <- showResultDolfinariumText$text
  getText1 <- iconv(as.character(getText1), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  dolfinarium.score <- suppressMessages(score.sentiment(getText1, pos.words, neg.words))
  
  hist(main="Sentiment Dolfinarium (april 4th 2013)", dolfinarium.score$score, breaks = (-10:10 + 0.7))
}

tweetsPathe1 <- function()
{
  
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  patheResultText <- dbSendQuery(doConnectToDB, "select text from tweetsPathe2")
  showResultPatheText <- fetch(patheResultText)
  onComplete <- dbHasCompleted(patheResultText)
  dbClearResult(patheResultText)
  dbDisconnect(doConnectToDB)
  
  
  getText2 <- showResultPatheText$text
  getText2 <- iconv(as.character(getText2), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  pathe.score <- suppressMessages(score.sentiment(getText2, pos.words, neg.words))
  
  hist(main="Sentiment Pathe (april 4th 2013)", pathe.score$score, breaks = (-10:10 + 0.7))
}

tweetsApenheul1 <- function()
{
  
  doConnectToDB <- dbConnect(MySQL(), user="root", password="", dbname="sicromoft", host="127.0.0.1")
  apenheulResultText <- dbSendQuery(doConnectToDB, "select text from tweetsApenheul2")
  showResultApenheulText <- fetch(apenheulResultText)
  onComplete <- dbHasCompleted(apenheulResultText)
  dbClearResult(apenheulResultText)
  dbDisconnect(doConnectToDB)
  
  
  getText3 <- showResultApenheulText$text
  getText3 <- iconv(as.character(getText3), "latin1", "ASCII", sub="")
  
  source("sentiment.R")
  source("readwords.R")
  
  apenheul.score <- suppressMessages(score.sentiment(getText3, pos.words, neg.words))
  
  hist(main="Sentiment Apenheul (april 4th 2013)", apenheul.score$score, breaks = (-10:10 + 0.7))
}

###################################################

freqEfteling <- function()
{
  tweets <- list()
  dates <- paste("2013-04-",11:16,sep="") 
  for (i in 2:length(dates)) {
    print(paste(dates[i-1], dates[i]))
    tweets <- c(tweets, searchTwitter("#efteling", since=dates[i-2], until=dates[i], n=170))
  }
  
  # Convert the list to a data frame
  tweets <- do.call("rbind", lapply(tweets, as.data.frame))
  
  #Extract some information
  tweets$date <- format(tweets$created, format="%Y-%m-%d")
  table(tweets$date)
  
  # Plot the frequency of tweets over time in two hour windows
  minutes <- 120
  x <- ggplot(data=tweets, aes(x=created)) + 
        geom_bar(aes(fill=..count..), binwidth=60*minutes) + 
        scale_x_datetime("Date") + 
        scale_y_continuous("Frequency") +
        ggtitle("#Efteling Tweet Frequency April 07-11")
  print(x)
}

freqPathe <- function()
{
  tweets <- list()
  dates <- paste("2013-04-",11:16,sep="") 
  for (i in 2:length(dates)) {
    print(paste(dates[i-1], dates[i]))
    tweets <- c(tweets, searchTwitter("#pathe", since=dates[i-2], until=dates[i], n=170))
  }
  
  tweets <- do.call("rbind", lapply(tweets, as.data.frame))

  tweets$date <- format(tweets$created, format="%Y-%m-%d")
  table(tweets$date)
  
  minutes <- 120
  x <- ggplot(data=tweets, aes(x=created)) + 
        geom_bar(aes(fill=..count..), binwidth=60*minutes) + 
        scale_x_datetime("Date") + 
        scale_y_continuous("Frequency") +
        ggtitle("#Pathe Tweet Frequency April 07-11")
  print(x)
}

freqDolfinarium <- function()
{
  tweets <- list()
  dates <- paste("2013-04-",11:16,sep="") 
  for (i in 2:length(dates)) {
    print(paste(dates[i-1], dates[i]))
    tweets <- c(tweets, searchTwitter("#dolfinarium", since=dates[i-2], until=dates[i], n=170))
  }
  
  tweets <- do.call("rbind", lapply(tweets, as.data.frame))
 
  tweets$date <- format(tweets$created, format="%Y-%m-%d")
  table(tweets$date)
  
  minutes <- 120
  x <- ggplot(data=tweets, aes(x=created)) + 
        geom_bar(aes(fill=..count..), binwidth=60*minutes) + 
        scale_x_datetime("Date") + 
        scale_y_continuous("Frequency") +
        ggtitle("#Dolfinarium Tweet Frequency April 07-11")
  print(x)
}

freqApenheul <- function()
{
  tweets <- list()
  dates <- paste("2013-04-",11:16,sep="") 
  for (i in 2:length(dates)) {
    print(paste(dates[i-1], dates[i]))
    tweets <- c(tweets, searchTwitter("#apenheul", since=dates[i-2], until=dates[i], n=150))
  }
  
  tweets <- do.call("rbind", lapply(tweets, as.data.frame))

  tweets$date <- format(tweets$created, format="%Y-%m-%d")
  table(tweets$date)
  
  minutes <- 120
  x <- ggplot(data=tweets, aes(x=created)) + 
        geom_bar(aes(fill=..count..), binwidth=60*minutes) + 
        scale_x_datetime("Date") + 
        scale_y_continuous("Frequency") +
        ggtitle("#Apenheul Tweet Frequency April 07-11")
  print(x)
}



# Define server logic here
shinyServer(function(input, output) {
  
  output$STPlot1 <- renderPlot({
    if(input$variable == "eft"){
      tweetsEfteling()
    }else if(input$variable == "dolf"){
      tweetsDolfinarium()
    }else if(input$variable == "pathe"){
      tweetsPathe()
    }else if(input$variable == "aap"){
      tweetsApenheul()
    }else{
      return(0)
    }
  })
  
  output$STPlot2 <- renderPlot({
    if(input$variable1 == "eft2"){
      tweetsEfteling1()
    }else if(input$variable1 == "dolf2"){
      tweetsDolfinarium1()
    }else if(input$variable1 == "pathe2"){
      tweetsPathe1()
    }else if(input$variable1 == "aap2"){
      tweetsApenheul1()
    }else{
      return(0)
    }
  })
  
  output$tweetFreq <- renderPlot({
    if(input$variable2 == "eft3"){
      freqEfteling()
    }else if(input$variable2 == "dolf3"){
      freqDolfinarium()
    }else if(input$variable2 == "pathe3"){
      freqPathe()
    }else if(input$variable2 == "aap3"){
      freqApenheul()
    }else{
      return(0)
    }
  })
  
})